### PR TITLE
Extract access tokens from HTTP requests

### PR DESCRIPTION
**Pull request by Kegsay**
Doesn't verify tokens yet.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/readers/sync.go [hunk 1]
```diff
@@ -3,13 +3,22 @@ package readers
...
 )
 
 // Sync implements /sync
 func Sync(req *http.Request) (interface{}, *util.HTTPError) {
 	logger := util.GetLogger(req.Context())
-	logger.Info("Doing stuff...")
+	userID, err := auth.VerifyAccessToken(req)
+	if err != nil {
+		return nil, &util.HTTPError{
+			Code: 403,
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/readers/sync.go]: I would very much like your thoughts on this. This has decoupled the error JSON and the HTTP status code. Should we be bundling them into one thing or?

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 2]
```diff
@@ -3,13 +3,26 @@ package writers
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/clientapi/auth"
 	"github.com/matrix-org/util"
 )
 
 // SendMessage implements /rooms/{roomID}/send/{eventType}
 func SendMessage(req *http.Request, roomID, eventType string) (interface{}, *util.HTTPError) {
 	logger := util.GetLogger(req.Context())
-	logger.WithField("roomID", roomID).WithField("eventType", eventType).Info("Doing stuff...")
+	userID, err := auth.VerifyAccessToken(req)
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Should we be doing this here or in `routing.go`? It feels like maybe `routing.go` but I'm unsure.

Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: From IRL: Keep it here.

File: src/github.com/matrix-org/dendrite/clientapi/readers/sync.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/readers/sync.go]: From IRL: Use `util.JSONResponse` instead and combine the concepts.

File: src/github.com/matrix-org/dendrite/clientapi/auth/auth.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+)
+
+// VerifyAccessToken verifies that an access token was supplied in the given HTTP request
+// and returns the user ID it corresponds to. Returns an error if there is no access token
+// or the token is invalid.
+func VerifyAccessToken(req *http.Request) (userID string, err error) {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: From IRL: Three-value return (user id, failure JSON response, everything-went-wrong error)

**Comment by Kegsay** [comment]
@NegativeMjark PTAL

File: src/github.com/matrix-org/dendrite/clientapi/auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"strings"
+
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/util"
+)
+
+// VerifyAccessToken verifies that an access token was supplied in the given HTTP request
+// and returns the user ID it corresponds to. Returns err if there was a fatal problem checking
+// the token. Returns resErr (an error response which can be sent to the client) if the token is invalid.
+func VerifyAccessToken(req *http.Request) (userID string, resErr *util.JSONResponse, err error) {
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: Hmm Now that I see it in the flesh the three value return is a bit cumbersome.

Maybe just return `(userID string, errResp *util.JsonResponse)` and wrap the `error`s as 500 inside `VerifyAccessToken`?

**Pull request merged**